File: android/src/test/java/com/margelo/nitro/nitrologger/FileSinkLifecycleRowsTest.kt
package com.margelo.nitro.nitrologger

import com.google.gson.JsonObject
import com.google.gson.JsonParser
import org.junit.After
import org.junit.Assert.assertEquals
import org.junit.Assert.assertNull
import org.junit.Assert.assertTrue
import org.junit.Assert.fail
import org.junit.Before
import org.junit.Test
import java.io.File

/**
 * The Android half of the shared no-handle table.
 *
 * The table lives in `spec/file-sink-lifecycle.rows.json` and is read by three
 * suites: this one, `__tests__/fileSinkLifecycleRows.test.ts`, and
 * `FileSinkLifecycleRowsTests.swift`. Its header explains why the answers live
 * in one file rather than in three hand-written suites that drifted apart on
 * four rows without anyone noticing.
 *
 * ## What this does NOT prove
 *
 * That `HybridFileSink` answers this way. That file extends a nitrogen-generated
 * base and cannot be constructed off a device, so nothing here executes a line
 * of it. [answer] below derives each answer from [FileSinkLifecycle] the way the
 * adapter does — deliberately through the same `snapshot()` / `artifactSource()`
 * calls, so a change to the lifecycle's verdict breaks this suite — but the
 * marshalling in between is out of reach until the `FileSinkAnswers` extraction
 * lands. At that point [answer] delegates instead of deriving, and these same
 * rows start pinning the adapter.
 */
class FileSinkLifecycleRowsTest {
  private lateinit var directory: File
  private lateinit var registry: LogWriterRegistry
  private val handles = mutableListOf<LogFileHandle>()

  @Before
  fun setUp() {
    directory = File.createTempFile("nitro-lifecycle-rows", "").let {
      it.delete()
      File(it.absolutePath + "-dir").apply { mkdirs() }
    }
    registry = LogWriterRegistry.isolated()
  }

  @After
  fun tearDown() {
    handles.forEach { runCatching { it.close(500.0) } }
    handles.clear()
    directory.deleteRecursively()
  }

  private fun handle(name: String = "app.log"): LogFileHandle = registry.acquire(
    path = File(directory, name).absolutePath,
    policy = LogRotationPolicy.of(),
    lineFramed = true,
    platform = PlatformIo.Jvm,
    clock = { 1_700_000_000_000L }
  ).also { handles.add(it) }

  // MARK: loading

  private data class Row(val op: String, val why: String, val answers: Map<String, Map<String, String>>)

  private data class Table(val modes: List<String>, val rows: List<Row>)

  private fun loadTable(): Table {
    // A JVM unit test's working directory is whatever the Gradle worker chose
    // and is not the repository, so `android/build.gradle` hands the root in.
    // A missing property is a failure rather than a default, because every
    // default here is a path that resolves to nothing and skips the suite.
    val root = System.getProperty("nitroLogger.repoRoot")
      ?: throw AssertionError(
        "nitroLogger.repoRoot is unset — see testOptions.unitTests in android/build.gradle"
      )
    val file = File(root, "spec/file-sink-lifecycle.rows.json")
    assertTrue("no table at ${file.absolutePath}", file.isFile)

    val json = JsonParser.parseString(file.readText()).asJsonObject
    val modes = json.getAsJsonArray("modes").map { it.asString }
    val rows = json.getAsJsonArray("rows").map { element ->
      val entry = element.asJsonObject
      val op = entry.get("op").asString
      Row(
        op = op,
        why = entry.get("why").asString,
        answers = modes.associateWith { mode ->
          val answer = entry.get(mode) as? JsonObject
            ?: throw AssertionError("row `$op` has no answer for mode `$mode`")
          answer.entrySet().associate { (key, value) -> key to value.asString }
        }
      )
    }
    return Table(modes, rows)
  }

  // MARK: the two modes

  /** A sink that was never opened: no handle, and nothing can exist yet. */
  private fun neverOpened(): FileSinkLifecycle = FileSinkLifecycle()

  /**
   * A sink that opened, wrote, and closed: no handle, and files may exist.
   *
   * The file is left on disk on purpose. `getLogFilePaths` enumerates the
   * directory rather than remembering a list, so a mode that deleted its file
   * would answer `pathCount: 0` and agree with the never-opened row for a
   * reason the table does not describe.
   */
  private fun openedThenClosed(): FileSinkLifecycle {
    val lifecycle = FileSinkLifecycle()
    assertEquals(FileSinkLifecycle.Claim.GRANTED, lifecycle.beginOpen())
    val live = handle()
    assertEquals(FileSinkLifecycle.Installation.INSTALLED, lifecycle.finishOpen(live))
    live.appendBatch("{\"m\":1}\n", 1L)
    lifecycle.beginClose().handle?.close(1000.0)
    assertNull("the mode is defined by having no handle", lifecycle.current())
    return lifecycle
  }

  private fun lifecycleFor(mode: String): FileSinkLifecycle = when (mode) {
    "neverOpened" -> neverOpened()
    "openedThenClosed" -> openedThenClosed()
    // Never a skip. A mode this target cannot build is a mode this target does
    // not test, and a green suite would report the opposite.
    else -> throw AssertionError("no builder for mode `$mode`")
  }

  // MARK: the dispatcher

  private val dispatchedOps = listOf(
    "appendBatch", "getStatus", "maintain", "collectLogs", "flush",
    "close", "clearLogs", "deleteSupportBundle", "getLogFilePaths"
  )

  /**
   * Each branch is the body `HybridFileSink` runs when its handle is null,
   * reading the same lifecycle calls in the same order.
   *
   * Returns null for an op this target does not implement, which the caller
   * turns into a failure.
   */
  private fun answer(op: String, lifecycle: FileSinkLifecycle): Map<String, String>? = when (op) {
    "appendBatch" -> {
      if (lifecycle.current() != null) null
      else mapOf(
        "accepted" to "false", "rejectReason" to "closed", "queuedBytes" to "0",
        "lostBytes" to "0", "lostEntries" to "0", "degraded" to "0"
      )
    }

    "getStatus", "maintain" -> {
      if (lifecycle.current() != null) null
      else mapOf(
        "queuedBytes" to "0", "lostBytes" to "0", "lostEntries" to "0", "degraded" to "0"
      )
    }

    "collectLogs" -> {
      if (lifecycle.current() != null) null
      else mapOf(
        "path" to "", "byteCount" to "0", "sourceFileCount" to "0",
        "complete" to "true", "truncated" to "false"
      )
    }

    "flush" -> {
      val snapshot = lifecycle.snapshot()
      if (snapshot.handle != null) null
      else noHandleFlush(snapshot.durableWithoutHandle)
    }

    "close" -> {
      val first = lifecycle.beginClose()
      if (first.handle != null) {
        null
      } else {
        val second = lifecycle.beginClose()
        assertEquals(
          "closing twice must answer what the first close answered",
          first.durableWithoutHandle, second.durableWithoutHandle
        )
        noHandleFlush(second.durableWithoutHandle)
      }
    }

    "clearLogs" -> {
      val snapshot = lifecycle.snapshot()
      if (snapshot.handle != null) null
      else mapOf(
        "deletedCount" to "0", "failedPathCount" to "0",
        "durable" to snapshot.durableWithoutHandle.toString(), "rebound" to "false"
      )
    }

    // `snapshot()`, not `artifactSource()`. A sink that opened and closed knows
    // where its bundle would be but cannot confirm it is gone, and `true` there
    // deletes the caller's obligation to retry. This is the one row that was
    // wrong in shipped code, and it was found by review rather than by a test
    // because no test could reach the file.
    "deleteSupportBundle" -> {
      val snapshot = lifecycle.snapshot()
      if (snapshot.handle != null) null
      else mapOf("deleted" to snapshot.durableWithoutHandle.toString())
    }

    "getLogFilePaths" -> {
      val source = lifecycle.artifactSource()
      when {
        source.handle != null -> null
        source.path == null -> mapOf("pathCount" to "0")
        else -> mapOf(
          "pathCount" to LogFileWriter.artifactPaths(File(source.path!!)).size.toString()
        )
      }
    }

    else -> null
  }

  private fun noHandleFlush(durable: Boolean): Map<String, String> = mapOf(
    "durable" to durable.toString(), "timedOut" to "false", "pendingBytes" to "0",
    "queuedBytes" to "0", "lostBytes" to "0", "lostEntries" to "0", "degraded" to "0"
  )

  // MARK: guards
  //
  // A shared table is only shared if every consumer is forced to keep up with
  // it. Each guard turns a way of quietly falling behind into a failure: a row
  // nobody dispatches, an op the table does not carry, a table that shrank.
  // Without them a new row would pass here by being ignored, and "add a row and
  // watch three suites go red" would describe something that does not happen.

  /**
   * Pinned, not derived. `rows.size >= rows.size` is the shape of gate this
   * repository has already shipped twice by accident.
   */
  private val pinnedRowFloor = 9
  private val pinnedModes = listOf("neverOpened", "openedThenClosed")

  @Test
  fun `the table still has every row this floor was pinned against`() {
    assertTrue(loadTable().rows.size >= pinnedRowFloor)
  }

  @Test
  fun `the table declares exactly the modes this target can build`() {
    assertEquals(pinnedModes, loadTable().modes)
  }

  @Test
  fun `every row names an op this target dispatches`() {
    val unknown = loadTable().rows.map { it.op }.filterNot { it in dispatchedOps }
    assertEquals("unimplemented rows are a failure, never a skip", emptyList<String>(), unknown)
  }

  @Test
  fun `every op this target dispatches has a row`() {
    val covered = loadTable().rows.map { it.op }.toSet()
    assertEquals(
      "an op answered here but absent from the table is answered by nobody else",
      emptyList<String>(), dispatchedOps.filterNot { it in covered }
    )
  }

  @Test
  fun `no op is listed twice`() {
    // Without this the headline claim — add a row, watch three suites go red —
    // has a hole in it: a tenth row duplicating an existing op and its answers
    // dispatches the same code a second time and passes everywhere. The row
    // count would even rise, satisfying every floor.
    val ops = loadTable().rows.map { it.op }
    assertEquals("a duplicated op tests nothing twice", ops.toSet().size, ops.size)
  }

  @Test
  fun `every row explains itself`() {
    // Not decoration. `complete: true` over an empty collect and
    // `durable: true` over a sink that never opened both read as bugs until you
    // know why.
    loadTable().rows.forEach {
      assertTrue("row `${it.op}` does not say why", it.why.length > 40)
    }
  }

  // MARK: the rows

  @Test
  fun `every row answers as the table says`() {
    val table = loadTable()
    var checked = 0

    for (mode in table.modes) {
      for (row in table.rows) {
        val expected = row.answers[mode]
        if (expected == null) {
          fail("row `${row.op}` has no answer for `$mode`")
          continue
        }
        val actual = answer(row.op, lifecycleFor(mode))
        if (actual == null) {
          fail("no dispatcher for `${row.op}`, or it found a live handle in `$mode`")
          continue
        }

        // Field-for-field in both directions, between the TABLE and the
        // DISPATCHER: a field the table names that [answer] does not produce
        // fails, and so does the reverse. It is deliberately not a claim about
        // SinkStatus or FlushOutcome growing a field — those types are not
        // reachable from this target at all, which is the whole reason this
        // file exists. The Jest half pins their key sets.
        assertEquals("`${row.op}` in mode `$mode`: ${row.why}", expected, actual)
        checked++
      }
    }

    // The loop above is only worth its assertions if it ran. A table that
    // parsed to zero rows, or a mode list that came back empty, would otherwise
    // report a pass having compared nothing.
    assertEquals(
      "every row must be checked in every mode",
      table.modes.size * table.rows.size, checked
    )
    assertTrue(checked >= pinnedModes.size * pinnedRowFloor)
  }
}

File: __tests__/fileSinkLifecycleRows.test.ts
import rows from '../spec/file-sink-lifecycle.rows.json';
import { MemoryWriter } from './helpers/MemoryFileSink';
import type { MemoryFileSink } from './helpers/MemoryFileSink';

/**
 * `MemoryFileSink` against the shared no-handle table.
 *
 * The table lives in `spec/file-sink-lifecycle.rows.json` and is read by three
 * suites: this one, `FileSinkLifecycleRowsTests.swift`, and
 * `FileSinkLifecycleRowsTest.kt`. Its header explains why the answers live in
 * one file rather than in three hand-written suites that drifted apart.
 *
 * Replaces `describe('MemoryFileSink — the no-handle rows both adapters
 * implement')`, which asserted the same answers by hand and could therefore
 * agree with a table nobody else was reading.
 *
 * ## What this file does NOT prove
 *
 * That the natives answer this way. Nothing in a Jest process can execute
 * Swift or Kotlin. It proves the double answers the way the table says, which
 * matters because the double is what every `FileDestination` test runs
 * against: a double that answers differently from the shipped adapters makes
 * every one of those tests a statement about nothing.
 */

/**
 * The field set each spec type actually has, pinned.
 *
 * Every dispatcher below builds a *projection* of its result, so on its own the
 * row comparison proves only that the table and the projection agree. A field
 * added to `SinkStatus` or `FlushOutcome` would be dropped before the
 * comparison and go on being unasserted by the table indefinitely — in a
 * release that is itself adding a method to this spec. Pinning the raw keys
 * makes that a failure here, which is the one suite whose job is that these
 * answers are complete.
 *
 * `deleteSupportBundle` and `getLogFilePaths` are absent because their results
 * are a boolean and an array: there are no keys to pin, and their single row
 * field is a derived count rather than a projection.
 */
const RAW_KEYS: Record<string, readonly string[]> = {
  appendBatch: [
    'accepted',
    'degraded',
    'lostBytes',
    'lostEntries',
    'queuedBytes',
    'rejectReason',
  ],
  getStatus: ['degraded', 'lostBytes', 'lostEntries', 'queuedBytes'],
  maintain: ['degraded', 'lostBytes', 'lostEntries', 'queuedBytes'],
  collectLogs: [
    'byteCount',
    'complete',
    'path',
    'sourceFileCount',
    'truncated',
  ],
  flush: [
    'degraded',
    'durable',
    'lostBytes',
    'lostEntries',
    'pendingBytes',
    'queuedBytes',
    'timedOut',
  ],
  close: [
    'degraded',
    'durable',
    'lostBytes',
    'lostEntries',
    'pendingBytes',
    'queuedBytes',
    'timedOut',
  ],
  clearLogs: ['deletedCount', 'durable', 'failedPaths', 'rebound'],
};

/** Fails if the result grew or lost a field, before it is projected away. */
function pin<T extends object>(op: string, raw: T): T {
  const expected = RAW_KEYS[op];
  if (expected) {
    expect(Object.keys(raw).sort()).toEqual([...expected].sort());
  }
  return raw;
}

const DISPATCH: Record<
  string,
  (sink: MemoryFileSink) => Record<string, string>
> = {
  appendBatch: (sink) => {
    const r = pin('appendBatch', sink.appendBatch('{"m":2}\n', 1));
    return {
      accepted: String(r.accepted),
      // `??` and not `||`: an implementation that answered `''` here would be
      // reporting "refused, and I decline to say why", which is not the row.
      rejectReason: String(r.rejectReason ?? '<absent>'),
      queuedBytes: String(r.queuedBytes),
      lostBytes: String(r.lostBytes),
      lostEntries: String(r.lostEntries),
      degraded: String(r.degraded),
    };
  },
  getStatus: (sink) => status(pin('getStatus', sink.getStatus())),
  maintain: (sink) => {
    const answer = status(pin('maintain', sink.maintain(1000)));
    // The sweep must not run. Asserted here rather than in the table because
    // it is a fact about the *call*, not about the answer, and the table's
    // shape is answer fields only.
    expect(sink.maintainCalls).toEqual([1000]);
    return answer;
  },
  collectLogs: (sink) => {
    const o = pin('collectLogs', sink.collectLogs(1000, 1_000_000));
    return {
      path: o.path,
      byteCount: String(o.byteCount),
      sourceFileCount: String(o.sourceFileCount),
      complete: String(o.complete),
      truncated: String(o.truncated),
    };
  },
  flush: (sink) => flush(pin('flush', sink.flush(1000))),
  close: (sink) => {
    const first = sink.close(1000);
    const second = sink.close(1000);
    // Idempotence is a relation between two calls, so it cannot be a row.
    // The table pins what the answer *is*; this pins that asking twice does
    // not change it.
    expect(second).toEqual(first);
    return flush(pin('close', second));
  },
  clearLogs: (sink) => {
    const o = pin('clearLogs', sink.clearLogs(1000));
    return {
      deletedCount: String(o.deletedCount),
      failedPathCount: String(o.failedPaths.length),
      durable: String(o.durable),
      rebound: String(o.rebound),
    };
  },
  deleteSupportBundle: (sink) => ({
    deleted: String(sink.deleteSupportBundle(1000)),
  }),
  getLogFilePaths: (sink) => ({
    pathCount: String(sink.getLogFilePaths().length),
  }),
};

function status(s: {
  queuedBytes: number;
  lostBytes: number;
  lostEntries: number;
  degraded: number;
}): Record<string, string> {
  return {
    queuedBytes: String(s.queuedBytes),
    lostBytes: String(s.lostBytes),
    lostEntries: String(s.lostEntries),
    degraded: String(s.degraded),
  };
}

function flush(o: {
  durable: boolean;
  timedOut: boolean;
  pendingBytes: number;
  queuedBytes: number;
  lostBytes: number;
  lostEntries: number;
  degraded: number;
}): Record<string, string> {
  return {
    durable: String(o.durable),
    timedOut: String(o.timedOut),
    pendingBytes: String(o.pendingBytes),
    ...status(o),
  };
}

/** A sink that was never opened: no handle, and nothing can exist yet. */
function neverOpened(): MemoryFileSink {
  return new MemoryWriter().attach();
}

/** A sink that opened, wrote, and closed: no handle, and files may exist. */
function openedThenClosed(): MemoryFileSink {
  const sink = new MemoryWriter().attach();
  sink.open('/memory/logs/app.log', undefined, true);
  sink.appendBatch('{"m":1}\n', 1);
  sink.close(1000);
  return sink;
}

const MAKE: Record<string, () => MemoryFileSink> = {
  neverOpened,
  openedThenClosed,
};

/**
 * The guards, which are the reason this file is worth more than the hand-
 * written suite it replaces.
 *
 * A shared table is only shared if every consumer is forced to keep up with
 * it. Each assertion below turns a way of quietly falling behind into a
 * failure: a row nobody dispatches, a mode nobody builds, a field the table
 * does not name, a table that shrank. Without them a new row would pass here
 * by being ignored, and "add a row and watch three suites go red" would be a
 * claim about a thing that does not happen.
 *
 * The floor is pinned rather than derived. `rows.length >= rows.length` is
 * the shape of gate this repository has already shipped twice by accident.
 */
const PINNED_ROW_FLOOR = 9;
const PINNED_MODES = ['neverOpened', 'openedThenClosed'];

describe('the shared no-handle table — guards', () => {
  test('has at least the rows it had when this floor was pinned', () => {
    expect(rows.rows.length).toBeGreaterThanOrEqual(PINNED_ROW_FLOOR);
  });

  test('declares exactly the modes this suite can build', () => {
    expect(rows.modes).toEqual(PINNED_MODES);
    expect(Object.keys(MAKE).sort()).toEqual([...PINNED_MODES].sort());
  });

  test('every row names an op this suite dispatches', () => {
    // A failure, never a skip. A skip here is how a row gets added to the
    // table, ignored by one target, and believed by everyone reading the file.
    const unknown = rows.rows
      .map((r) => r.op)
      .filter((op) => !(op in DISPATCH));
    expect(unknown).toEqual([]);
  });

  test('every op this suite dispatches has a row', () => {
    const covered = new Set(rows.rows.map((r) => r.op));
    expect(Object.keys(DISPATCH).filter((op) => !covered.has(op))).toEqual([]);
  });

  test('no op is listed twice', () => {
    // Without this the headline claim — add a row, watch three suites go red —
    // has a hole in it: a tenth row duplicating an existing op and its answers
    // dispatches the same code a second time and passes everywhere. The row
    // count would even rise, satisfying every floor.
    const ops = rows.rows.map((r) => r.op);
    const duplicated = ops.filter((op, i) => ops.indexOf(op) !== i);
    expect(duplicated).toEqual([]);
  });

  test('every row answers for every declared mode', () => {
    for (const row of rows.rows) {
      const entry = row as unknown as Record<string, unknown>;
      for (const mode of rows.modes) {
        expect(typeof entry[mode]).toBe('object');
      }
    }
  });

  test('every row explains itself', () => {
    // Not decoration. The answers here are not derivable from the type
    // signatures — `complete: true` over an empty collect and `durable: true`
    // over a sink that never opened both read as bugs until you know why.
    for (const row of rows.rows) {
      expect(row.why.length).toBeGreaterThan(40);
    }
  });
});

describe.each(PINNED_MODES)('MemoryFileSink — %s', (mode) => {
  const make = MAKE[mode]!;

  test.each(rows.rows.map((r) => [r.op, r] as const))('%s', (op, row) => {
    const expected = (row as unknown as Record<string, Record<string, string>>)[
      mode
    ]!;
    const actual = DISPATCH[op]!(make());

    // Field-for-field in both directions, between the TABLE and the DISPATCHER:
    // a field the table names that the dispatcher does not produce fails, and
    // so does the reverse. It is deliberately not a claim about the sink's own
    // type growing a field — the dispatcher projects, so an extra field would
    // be discarded before it got here. `RAW_KEYS` is what covers that.
    expect(actual).toEqual(expected);
  });
});
